packages/pay/src/utils.ts

```diff
@@ -129,48 +47,52 @@ export class Int {
     this.value = n
   }
 
+  static from<T extends Int>(n: T): T
   static from(n: Long): Int
+  static from(n: NonNegativeInteger): Int
   static from(n: number): Int | undefined
+  static from(n: bigint): Int | undefined
   static from(n: string): Int | undefined
-  static from(n: BigNumber): Int | undefined
-  static from(n: Long | number | string | BigNumber): Int | undefined {
-    if (typeof n === 'string') {
+  static from(n: Int | bigint | number | string): Int | undefined // Necessary for amounts passed during setup
+  static from<T extends Long | Int | bigint | number | string>(n: T): Int | undefined {
+    if (n instanceof Int) {
+      return new Int(n.value)
+    } else if (typeof n === 'bigint') {
+      return Int.fromBigint(n)
+    } else if (typeof n === 'string') {
       return Int.fromString(n)
     } else if (typeof n === 'number') {
       return Int.fromNumber(n)
-    } else if (BigNumber.isBigNumber(n)) {
-      return Int.fromBigNumber(n)
-    } else {
+    } else if (Long.isLong(n)) {
       return Int.fromLong(n)
     }
   }
 
+  private static fromBigint(n: bigint): Int | undefined {
+    if (n >= 0) {
+      return new Int(n)
+    }
+  }
+
   private static fromString(n: string): Int | undefined {
     try {
-      const big = BigInt(n)
-      if (big >= 0) {
-        return new Int(big)
-      }
+      return Int.fromBigint(BigInt(n))
       // eslint-disable-next-line no-empty
     } catch (_) {}
```

Comment:
It feels weird to me that we're swallowing errors everywhere and returning undefined. It seems like a good way for undefined and NaN to leak into our calculations. Are we checking that the result is defined everywhere we use this methods?

Reply:
> Are we checking that the result is defined everywhere we use this methods?

TypeScript enforces that the consumer checks it's defined before accessing any method, so the consumer must narrow the type by implementing a code path for the error case. For example, here:

https://github.com/interledgerjs/interledgerjs/blob/cc286cea8e17380bc4a7db351cc45209d2bf43fe/packages/pay/src/index.ts#L250-L252

A big benefit to this abstraction is it prevents `NaN`, `Infinity`, etc., since all validation on numbers is done before instantiating an `Int` or a `Ratio`, and their internal logic prevents this. So, coercing it back to a number later is always safe. `Int` should not be used with arithmetic operators, but if it is, TS would error given something like `Int.ONE + 1`.

Assertions or casting could result in this, but that's generally an antipattern and checked against with linting (we do this several places in the codebase, but only when creating Ints from constants).